File: Aula_Ricardo/Programação_orientada_a_objetos/python/Interfaces_protocolos.py
from typing  import Protocol

class Imprimivel(Protocol):
    def imprimir(self) -> None:
        ...

class Relatorio:
    def imprimir(self) -> None:
        print("Imprimindo relatorio.")

class Contrato:
    def imprimir(self) -> None:
        print("Imprimindo contrato.")

def processar_impressao(documento: Imprimivel) -> None:
    documento.imprimir()

relatorio = Relatorio()
contrato = Contrato()

processar_impressao(relatorio)
processar_impressao(contrato)

File: Aula_Ricardo/Programação_orientada_a_objetos/java/Interfaces_protocolos.java
interface Imprimivel {
    void imprimir();
}

class Relatorio implements Imprimivel {
    @Override
    public void imprimir() {
        System.out.println("Imprimindo relatorio.");
    }
}

class Contrato implements Imprimivel {
    @Override
    public void imprimir() {
        System.out.println("Imprimindo contrato.");
    }
}

public class Main {
    public static void processarImpressao(Imprimivel documento) {
        documento.imprimir();
    }

    public static void main(String[] args) {
        Imprimivel relatorio = new Relatorio();
        Imprimivel contrato = new Contrato();

        processarImpressao(relatorio);
        processarImpressao(contrato);
    }
}
